### ISSUE TITLE
Unify Feature Transformations and Feature Views

### ISSUE THREAD

**Issue by franciscojavierarceo** [feast-dev/feast#4584]
**Is your feature request related to a problem? Please describe.**

In Feast, the current On Demand Feature View executes feature transformations at read time. However, this behavior is not immediately obvious from the name "On Demand Feature View." The term "On Demand" doesn't explicitly convey when the transformation occurs, leading to potential confusion for users trying to understand the timing and execution flow of feature transformations. Additionally, there's inconsistency in how different types of feature views (batch, streaming, on-demand) are declared and used, which can make the codebase less intuitive and harder to maintain.

**Describe the solution you'd like**
I propose to refactor and rename the way transformations are defined in Feast to make the execution timing explicit. Introducing a @transform decorator with a type parameter that accepts a FeastTransformation enum can achieve this clarity. The enum would define the transformation types:

```python
from enum import Enum

class FeastTransformation(Enum):
    ON_READ = 1      # Equivalent to the current On Demand Feature View
    ON_WRITE = 2     # New: Transformations executed at write time
    BATCH = 3        # New: Batch processing transformations
    STREAMING = 4    # Migrates the existing stream_feature_view decorator

@transform(
    type=FeastTransformation.ON_WRITE,
    schema=[...],
    entity=[...],
    sources=[...],
)
def my_feature_view(inputs: List[Dict[str, float]]) -> List[Dict[str, float]]:
    # Transformation logic here
    return outputs
```

By explicitly declaring when the transformation should occur (`ON_READ`, `ON_WRITE`, `BATCH`, or `STREAMING`), the code becomes more readable and maintainable. 

This approach unifies the declaration of feature views and transformations, making it easier for users to understand and for developers to extend functionality in the future.

**Describe alternatives you've considered**

* Enhancing Documentation: Improving the existing documentation to better explain the behavior of On Demand Feature View without changing the code structure. However, this doesn't solve the inherent ambiguity in the naming.

* Renaming Existing Decorators: Simply renaming on_demand_feature_view to something like on_read_feature_view to make the execution timing clearer. While this improves clarity, it doesn't address the inconsistency in how different feature views are declared.
* Using Configuration Parameters: Adding parameters to existing decorators to specify execution timing. This could make the decorators more complex and doesn't provide a unified approach.

**Additional context**
Consistency and Extensibility: This change promotes a consistent method of declaring feature transformations, regardless of when they execute. It also sets a foundation for adding more transformation types in the future.

Clarity for New Users: Making the execution timing explicit in the code helps new users understand the flow of data and transformations in Feast, reducing the learning curve.

Impact on Existing Codebase: Careful consideration and planning would be needed to migrate existing code to the new structure without breaking functionality. Deprecation warnings and migration guides could facilitate this transition.

Community Feedback: Engaging with the Feast community to gather feedback on this proposed change could provide additional insights and help refine the solution.

**Comment by franciscojavierarceo** [comment]
Fyi @tokoko @HaoXuAI @shuchu

**Comment by tokoko** [comment]
@franciscojavierarceo thanks. I like the simplicity of it, but wouldn't this be a problem when different types of FVs require different parameters? We can enumerate all the differences, but the first one that comes to mind is that in case of BatchFeatureViews `sources` is actually a list of `DataSource` objects, but for OnDemandFeatureViews `sources` is a list of `FeatureView` or `RequestSource` objects. We might end up with a less user-friendly interface, because we will no longer be able to convey this kind of information through typing.

I'm at the moment leaning more towards "better documentation + better naming" alternative because of these reasons.

**Comment by franciscojavierarceo** [comment]
>We can enumerate all the differences, but the first one that comes to mind is that in case of BatchFeatureViews sources is actually a list of DataSource objects, but for OnDemandFeatureViews sources is a list of FeatureView or RequestSource objects.

This is just a DAG and we can express this thoughtfully and it ends up having a lot of pros, much like how [dbt](https://docs.getdbt.com/terms/dag#dag-use-cases-and-best-practices) can compile a visual DAG to highlight behavior.

> We might end up with a less user-friendly interface, because we will no longer be able to convey this kind of information through typing.

I think this can be thought through. 

>I'm at the moment leaning more towards "better documentation + better naming" alternative because of these reasons.

I think the long term approach of making it clear to data scientists and engineers what is actually happening is worth this effort. Making users learn weird jargon instead of using obvious industry patterns is a confusing choice.

**Comment by franciscojavierarceo** [comment]
This means the following implementations will have to happen:

- [ ] Rename and migrate OnDemandFeatureView
- [ ] Rename and migrate StreamFeatureView
- [ ] Implement the batch transformation as a subset of `transform`
- [ ] Update lots of documentation

**Comment by HaoXuAI** [comment]
Hey I know you really want to optimize the ondemandfeatureciew, but I would recommend to implement the batch and stream feature view with transformation first, then come back to refactor and optimization. We need more use case and experiment to support the refactoring

**Comment by franciscojavierarceo** [comment]
Yes, let me clarify. I intend on doing the following:

1. Finish implementing transformations on writes (what I'm doing now)
2. Implement Batch transformations (next)
3. Rename and migrate Stream, On Demand, and Batch Feature Views

I should have ordered the above that way. So the full refactor and optimization is last.

**Comment by HaoXuAI** [comment]
that sounds good. i can help with the batch or stream transformation as well.

**Comment by franciscojavierarceo** [comment]
Sounds good. Want to take on the batch transformation?

You can tag me in the PR if u want feedback/help

**Comment by franciscojavierarceo** [comment]
There's already a spec written and maybe some code in place (take a look at aggregations.py) but I'm not sure what the status was.

**Comment by HaoXuAI** [comment]
Sure, I can play with it and share some thoughts, gonna finish some left work for the vector db

**Comment by dandawg** [comment]
I'm really glad to see the effort towards clarity in the API. I like the idea of getting rid of "on-demand". Can we expand on the idea of "when" to include "where/how" a transformation occurs? I've seen the pattern of "source" and "destination" in other transformation utilities (I know GCP and I think AWS uses this). 

As a data scientist, I often choose to have transformations occur in specific locations (this is just my DS work outside of Feast). For example, I may do initial transformations in-database (server-side), and then pull the transformed data client side (this is great for aggregations). Or, I may choose to pull the data first, and then transform client-side, or send it to a remote job (on say a spark cluster), and pull the output from there. Or, I may just have the output written back to the original DB.

Given these sorts of complexities, I'd love to have some flexibility to make these decisions in a Feast transformation. There are a few ways we could tackle the "where". One would be through separate methods, something like:
```
db_transform(source=table_a, destination=table_b)  #in-db transformation
spark_transform(source=table_a, destination=..., spark_endpoint=...)

```
Or, perhaps even more simplified:
```
transform(source=..., dest=..., config=...) 
# config could include info like 'spark engine' etc.
```
^ these ideas are just brainstorming--they would obviously need to be fleshed out more.

**Comment by franciscojavierarceo** [comment]
>Can we expand on the idea of "when" to include "where/how" a transformation occurs? I've seen the pattern of "source" and "destination" in other transformation utilities (I know GCP and I think AWS uses this).

Ultimately this maps to the "engine" that's used for transformation mentioned briefly in this document: https://docs.feast.dev/master/getting-started/architecture/feature-transformation#feature-transformation-engines

We have to think through how we orchestrate the transformation because Feast itself is not an orchestrator. I think Kubeflow Pipelines is a natural fit here and we could offer tutorials using Airflow as well.

So the `transform` decorator really just needs to store the transformation that can be executed on the engine and, thus, the task becomes to ensure that the decorator is compatible with the engine.

**Comment by tokoko** [comment]
> I should have ordered the above that way. So the full refactor and optimization is last.

+1 on this, let's get all the core functionality in and then decide on the interface.

P.S. I'm thinking of taking on #4277, I think no one has started on it yet, right?

**Comment by franciscojavierarceo** [comment]
Yeah, that sounds great! Do you want to tag me early in the PR so we can make sure we're on the same page for the refactor? I imagine we'll end up coming up with the same conclusion but would be good to collaborate on it.

**Comment by tokoko** [comment]
sure, I will start a draft PR here

**Comment by robhowley** [comment]
This will be a really nice improvement on the api. As it stands it's currently quite unintuitive, but a `@transform` decorator is a very clear and straightforward. Good readability change.

**Comment by emgeee** [comment]
I think this is a great idea, the current naming is confusing. I also think a diagram would help clarify the things

**Comment by franciscojavierarceo** [comment]
The doc has been authored here: https://docs.google.com/document/d/1KXCXcsXq1bUvbSpfhnUjDSsu4HpuUZ5XiZoQyltCkvo/edit?tab=t.0#heading=h.g6j9jyjkf5sm

I'll be linking the GitHub PRs to help provide clarity to folks in the future and updating the doc as progress happens.

**Comment by rehevkor5** [comment]
What's the purpose of `NONE`? If there is no transformation to do, why would you have a `@transform` at all? Shouldn't the definition of sources/entity/schema be independent of the transformations (if there even are any)? It seems like the definition of those are their own thing, and these transform methods are just optional interceptors that you can add at your desired place in the end-to-end data flow.

What's the purpose of / difference between `ON_WRITE` as opposed to `BATCH` and `STREAMING`?

One thing that is arguably missing from Feast is that if the code of an ODFV changes, it can break training pipelines & inference clients who have a dependency on it: no versioning is being used to provide compatibility guarantees. I'm not sure if anyone else is concerned about that, but it seems like it's worth considering.

**Comment by franciscojavierarceo** [comment]
> What's the purpose of `NONE`? If there is no transformation to do, why would you have a `@transform` at all? Shouldn't the definition of sources/entity/schema be independent of the transformations (if there even are any)? It seems like the definition of those are their own thing, and these transform methods are just optional interceptors that you can add at your desired place in the end-to-end data flow.

Right, that would be the default when not in place. We could probably omit it, I left it for completeness.

> What's the purpose of / difference between `ON_WRITE` as opposed to `BATCH` and `STREAMING`?

Batch and streaming don't require a transformation on write, their respective "transformation engines" handle that for you. In batch land, that could be your offline store (e.g,. Spark, Snowflake, etc.) and in streaming that could be Spark Streaming of Flink. The `ON_WRITE` case is actually using the Python feature server as the transformation.


> One thing that is arguably missing from Feast is that if the code of an ODFV changes, it can break training pipelines & inference clients who have a dependency on it: no versioning is being used to provide compatibility guarantees. I'm not sure if anyone else is concerned about that, but it seems like it's worth considering.

This is a really great point! And you're absolutely right. This starts to get complicated and, in short, you have to manage some sort of dependency on the client version in the worst case. We could definitely add metadata to be returned in the response to allow the client to validate that it's the expected transformation version. 

Let me know if you were thinking of something else, if something's not clear, or if you have other thoughts! Appreciate the feedback!

**Comment by rehevkor5** [comment]
So... BATCH is applied what, when reading from the offline store?

And STREAMING is applied when using Feast to run a Spark/Flink job that writes into offline and/or online store?

> Batch and streaming don't require a transformation on write, their respective "transformation engines" handle that for you.

Sorry, this confuses me... they don't require it, sure... but Feast provides a way, if you want to use it, of writing a `@stream_feature_view` which passes your Python function a Spark DataFrame, right? And that Python function becomes a transformation that is being done at the time of writing into Feast's stores, right? That seems like a "transformation on write" to me.

So `ON_WRITE` stuff runs within the Python feature server, ok. But what kind of write are we talking about? Materialization? (If so then it seems like it would be better named `ON_MATERIALIZE`).

Minor note: instead of "BATCH" and "STREAMING", more consistent to say "BATCH" and "STREAM".

**Comment by franciscojavierarceo** [comment]
FYI @ntkathole we should incorporate feedback on this ticket too: https://github.com/feast-dev/feast/issues/4277#issuecomment-2427772928
